metrics/thresholds.go

```diff
@@ -244,6 +248,58 @@ func (ts *Thresholds) Parse() error {
 	return nil
 }
 
+// ErrInvalidThreshold indicates a threshold is not valid
+var ErrInvalidThreshold = errors.New("invalid threshold")
+
+// Validate ensures a threshold definition is consistent with the metric it applies to.
+// Given a metric registry and a metric name to apply the expressions too, Validate will
+// assert that each threshold expression uses an aggregation method that's supported by the
+// provided metric. It returns an error otherwise.
+// Note that this function expects the passed in thresholds to have been parsed already, and
+// have their Parsed (ThresholdExpression) field already filled.
+func (ts *Thresholds) Validate(metricName string, r *Registry) error {
+	parsedMetricName, _, err := ParseMetricName(metricName)
+	if err != nil {
+		err := fmt.Errorf("unable to validate threshold expressions; reason: %w", ErrMetricNameParsing)
+		return errext.WithExitCodeIfNone(err, exitcodes.InvalidConfig)
+	}
+
+	// Obtain the metric the thresholds apply to from the registry.
+	// if the metric doesn't exist, then we return an error indicating
+	// the InvalidConfig exitcode should be used.
+	metric := r.Get(parsedMetricName)
+	if metric == nil {
+		err := fmt.Errorf("%w defined on %s; reason: no metric name %q found", ErrInvalidThreshold, metricName, metricName)
+		return errext.WithExitCodeIfNone(err, exitcodes.InvalidConfig)
+	}
+
+	for _, threshold := range ts.Thresholds {
+		// Return a digestable error if we attempt to validate a threshold
+		// that hasn't been parsed yet.
+		if threshold.parsed == nil {
```

Comment:
Is there particular reason why we shouldn't parse it here?

Reply:
You mean that we should parse it if we find it to be nil? I must admit that I didn't think of doing that, but that's a good idea, will do so 👍🏻

Reply:
There might be some reason not to do it here- I haven't looked at how this gets called. But if we can that might make using this function easier

Reply:
I've given some thought to it, and I believe it's okay to do it. `Parse` and `Validate` are two different thresholds, but if one calls `Validate` with a threshold unparsed, it will be a good sentinel to try to parse it anyway. The updated code also makes sure to return the resulting parse error 👍🏻